rfd/0156-notifications.md

```diff
@@ -0,0 +1,404 @@
+   // Code is the unique notification code.
+   string Code = 2 [(gogoproto.jsontag) = "code"];
+   // Title is the title/header of the notification.
+   string Title = 3 [(gogoproto.jsontag) = "title"];
+   // Description is the text content of the notification.
+   string Description = 4 [(gogoproto.jsontag) = "description"];
+   // Created is when the notification was created, in UNIX time.
+   google.protobuf.Timestamp Created = 5 [(gogoproto.jsontag) = "created"];
+   // TTL is how long the notification should last for before automatically expiring. Default is 30 days.
+   int64 TTL = 6 [(gogoproto.jsontag) = "ttl,omitempty"];
+   // Metadata is optional additional metadata to be added to the notification.
+   google.protobuf.Struct Metadata = 7 [(gogoproto.jsontag) = "metadata,omitempty"];
+}
+```
+
+###### User Notification Preferences Item
+
+Each user will have one `UserNotificationStates` object which will keep track of the notifications they have seen or dismissed.
+
+These will be stored once per user in the database under `/notifications/states/<username>`
+
+```protobuf
+// UserNotificationStates represents a user's notification states. This is to keep track
+// of the notifications the user has seen and dismissed.
+message UserNotificationStates {
+   // ReadNotifications is a list of notification IDs that the user has already seen and will be marked as read.
+   repeated string ReadNotifications = 1 [(gogoproto.jsontag) = "read_notifications"];
+   // DismissedNotifications is a list of notification IDs that the user has chosen to hide. These will be filtered out when notifications are fetched.
+   repeated string DismissedNotifications = 2 [(gogoproto.jsontag) = "dismissed_notifications"];
+}
+```
+
+###### Global Notifications
+
+These will be stored under `/notification/global/<notification id>`.
+
+```protobuf
+// GlobalNotification represents a global notification.
+message GlobalNotification {
+   // Matcher for determining the target of this notification.
+   oneof Matcher {
+   // ByPermissions represents the RoleConditions needed for a user to receive this notification.
+   // If multiple permissions are defined and `MatchAllConditions` is true, the user will need to have
+   // all of them to receive this notification.
+   repeated RoleConditions ByPermissions = 1 [(gogoproto.jsontag) = "by_permissions"];
+   // ByRoles represents the roles targeted by this notification.
+   // If multiple roles are defined and `MatchAllConditions` is true, the user will need to have all
+   // of them to receive this notification.
+   repeated string ByRoles = 2 [(gogoproto.jsontag) = "by_roles"];
+   // All represents whether to target all users, regardless of roles or permissions.
+   bool All = 3 [(gogoproto.jsontag) = "all"];
+   }
+   // MatchAllConditions is whether or not all the conditions specified by the matcher must be met,
+   // if false, only one of the conditions needs to be met.
```

Comment:
We want to be able to target both use cases, and we wouldn't be able to target specific roles by passing in `RoleConditions`, only permissions. The implementation for the roles matcher would be quite simple anyways, since all it would involve is comparing the specified `ByRoles` strings with the roles the user has (which can be obtained from `user.GetRoles()`).